lib/ansible/modules/windows/win_domain_group_membership.ps1

```diff
@@ -0,0 +1,125 @@
+    $extra_args.Credential = $credential
+}
+if ($domain_server -ne $null) {
+    $extra_args.Server = $domain_server
+}
+
+$result = @{
+    changed = $false
+}
+if ($state -in @("present", "pure")) {
+    $result.added = @()
+}
+if ($state -in @("absent", "pure")) {
+    $result.removed = @()
+}
+if ($diff_mode) {
+    $result.diff = @{}
+}
+
+$members_before = Get-AdGroupMember -Identity $name @extra_args
+$pure_members = @()
+
+foreach ($member in $members) {
+    $group_member = Get-ADObject -Filter "SamAccountName -eq '$member' -and (ObjectClass -eq 'user' -or ObjectClass -eq 'group' -or ObjectClass -eq 'computer' -or ObjectClass -eq 'msDS-ManagedServiceAccount')"
+    if (!$group_member) {
+        Fail-Json -obj $result "Could not find domain user, group, service account or computer named $member"
+    }
+
+    if ($state -eq "pure") {
+        $pure_members += $group_member
+    }
+
+    $user_in_group = $false
+    foreach ($current_member in $members_before) {
+        if ($current_member.sid -eq $group_member.sid) {
+            $user_in_group = $true
+            break
+        }
+    }
+
+    if ($state -in @("present", "pure") -and !$user_in_group) {
+        Add-ADGroupMember -Identity $name -Members $group_member -WhatIf:$check_mode @extra_args
+        $result.added += $group_member.SamAccountName
+        $result.changed = $true
+    } elseif ($state -eq "absent" -and $user_in_group) {
+        Remove-ADGroupMember -Identity $name -Members $group_member -WhatIf:$check_mode @extra_args -Confirm:$False
+        $result.removed += $group_member.SamAccountName
+        $result.changed = $true
+    }
+}
+
+if ($state -eq "pure") {
+    # Perform removals for existing group members not defined in $members
+    $current_members = Get-AdGroupMember -Identity $name @extra_args
+
+    foreach ($current_member in $current_members) {
+        $user_to_remove = $true
+        foreach ($pure_member in $pure_members) {
+            if ($pure_member.sid -eq $current_member.sid) {
+                $user_to_remove = $false
+                break
+            }
+        }
+
+        if ($user_to_remove) {
+            Remove-ADGroupMember -Identity $name -Members $current_member -WhatIf:$check_mode @extra_args -Confirm:$False
+            $result.removed += $current_member.SamAccountName
+            $result.changed = $true
+        }
+    }
+}
+
+$final_members = Get-AdGroupMember -Identity $name @extra_args
+
+if ($final_members) {
+    $result.members = [Array]$final_members.SamAccountName
+} else {
+    $result.members = @()
+}
+
+if ($diff_mode -and $result.changed) {
+    $result.diff.before = $members_before.SamAccountName | Out-String
+    if (!$check_mode) {
+        $result.diff.after = [Array]$final_members.SamAccountName | Out-String
+    } else {
+        $after = [System.Collections.ArrayList][Array]$final_members.SamAccountName
```

Comment:
The use of `ArrayList` here means that `Add` is returning the index where the object was added at. This can cause issues when Ansible goes to parse the return json from a module. Unfortunately the Remove method of a List does return a bool but we can still handle that.

I would recommend you use a List like

```
$after = [System.Collections.Generic.List`1[String]]$final_members.sAMAccountName
```

There should also be no reason why you need the `[Array]` cast as well.